### PR TITLE
perf: Fuse sequence packing for loss function

### PR DESCRIPTION
**Pull request by nujoug**
# What does this PR do ?

Apply a single call of loss function to all sequences instead of calling for each individual sequences

# Issues
Issue #1247 

# Usage
Set flag `policy.sequence_packing.fuse_loss` to true to turn on this feature

# Results

Observed up to 15% speedup on policy training flops

W B Chart 3_10_2026, 9_28_09 PM

Validation results show similar accuracy curve

W B Chart 3_10_2026, 9_27_19 PM

# Additional Information
Check out this [report](https://wandb.ai/nujoug2-nvidia/nemo-rl-chunk-memory/reports/Sequence-Packing-Loss-Fusion--VmlldzoxNTkwMDg0OQ?accessToken=mtijty8rruimu0wv4hatw4cqwr1kmzbpn1kmr0quw6pjcle6vpwozx12tcw1mpk8) for more detailed analysis

## Summary by CodeRabbit

* **New Features**
  * Introduced fused sequence packing loss computation, enabling single-pass loss calculation instead of per-sequence processing. Users can now enable loss fusion via the `fuse_loss` configuration flag to improve training performance.

* **Tests**
  * Added comprehensive validation tests for fused loss wrapper across multiple distributed training configurations.

### CODE REVIEW

File: tests/unit/algorithms/test_sequence_packing_fusion.py [hunk 1]
```diff
@@ -0,0 +1,489 @@
...
+    Verifies that the fused wrapper produces identical:
+      - loss values
+      - backward gradients w.r.t. vocab-parallel logits
+    for different CP and TP configurations, and both with and without
+    pre-packed input_ids cached in the data dict.
+    """
+    cp_size, tp_size = cp_tp
+    cluster = cluster_fixture
+    actor_fqn = register_sequence_packing_loss_wrapper_baseline_actor
+    world_size = cp_size * tp_size
```
[the remark this answers is not in the document]
Comment (nujoug) [reply tests/unit/algorithms/test_sequence_packing_fusion.py:399]: done

File: tests/unit/algorithms/test_sequence_packing_fusion.py [hunk 2]
```diff
@@ -0,0 +1,464 @@
...
+                    :,
+                    run_seq // self.cp_size : (run_seq + padded_seq_len)
+                    // self.cp_size,
+                    :,
+                ] = _get_tokens_on_this_cp_rank(
+                    tmp, torch.distributed.get_rank(cp_group), self.cp_size
+                )
+                run_seq += padded_seq_len
+
+            return logits_local, packed_logits
```
[the remark this answers is not in the document]
Comment (nujoug) [reply tests/unit/algorithms/test_sequence_packing_fusion.py]: Added sanity check in test_sequence_packing_fusion.py line  448-451 to ensure the gradient is not None or all zeros.

```
# Sanity: gradients must be non-None and non-zero (autograd is connected)
for label in ("baseline", "candidate"):
    grad = r[label]["logits_local_grad"]
    assert grad is not None, f"{label} grad is None on rank {rank}"
    assert grad.abs().sum() > 0, f"{label} grad is all zeros on rank {rank}"
```

Unit test passed indicating that the gradient flow is correct

**Comment by nujoug** [comment]
@coderabbit review
